parsl/monitoring/db_manager.py

```diff
@@ -271,8 +273,11 @@ def start(self, priority_queue, resource_queue):
                         if msg['task_time_returned'] is not None:
                             update_messages.append(msg)
                         else:
-                            inserted_tasks.add(msg['task_id'])
-                            insert_messages.append(msg)
+                            if msg['task_id'] in inserted_tasks:
```

Comment:
I'm not clear what this change is trying to achieve: is it solving some problem related to duplicate inserts?

Reply:
For each retry of a task, monitoring receives a message of the task and try to insert the message again to db. However, the messages of the same task cannot be inserted again (identified as duplicate inserts due to duplicate primary IDs of task ID and workflow ID). This is to ensure that if the task is inserted before, we just do updates on several columns. Did I make it clear?

Reply:
Actually this is a bug but we did not test `retries` before -- it is not related to the PR title maybe

Reply:
If retries were broken in some cases before and this fixes it, this is worth putting in the PR title and description.

Reply:
Modified

Reply:
OK. So there's now a three way case here that I don't fully understand:

case 1:   `task_time_returned` is not None:    always update, never insert.
case 2:   `task_time_returned` is None, task_id is not in inserted tasks:  insert
case 3:   `task_time_returned` is None, task_id is in inserted tasks: update

I think case 1 here would be handled by the two new cases 2 and 3, and doesn't need to exist as a separate case;

and,

case 1 at the moment means that we must send a task message *without* a task_time_returned first, before reporting a task message *with* task_time_returned set. That is how the DFK seems to call things now (although I haven't verified properly) but it seems an unnecessary complication in this code.

Reply:
Oh you are right, good catch! It should be fixed now.